Segment writer fails to re-establish a connection post a network glitch.

**Problem description**
Segment writers are not able to establish a connection if the connection fails before the writer is able to send to the `SETUP_APPEND` wirecommand.
```
2020-07-30 14:23:16,635 616783 [clientInternal-2-1] WARN  i.p.c.s.i.SegmentOutputStreamImpl - 38fdccb6-27bd-4417-a767-ae9a90fecaae Failed to connect: 
java.util.concurrent.CompletionException: io.pravega.common.ObjectClosedException: Object 'TcpClientConnection' has been closed and cannot be accessed anymore.

2020-07-30 14:23:16,645 616793 [clientInternal-2-1] WARN  i.p.c.s.i.SegmentOutputStreamImpl - 38fdccb6-27bd-4417-a767-ae9a90fecaae Failed to connect: 
java.util.concurrent.CompletionException: java.lang.IllegalStateException
...	java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: null
	at com.google.common.base.Preconditions.checkState(Preconditions.java:492)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.lambda$null$4(SegmentOutputStreamImpl.java:584)
```

**Problem location**
Pravega client.

**Suggestions for an improvement**
EOFException can trigger this path 
`2020-07-30 14:23:16,578 616726 [ClientSocketReaders-244] INFO  i.p.c.c.impl.TcpClientConnection - Closing TcpClientConnection.Reader because end of input reached.`
[readwriteandscalewithfailovertest-h85hc-1.zip](https://github.com/pravega/pravega/files/5006313/readwriteandscalewithfailovertest-h85hc-1.zip)

--- a/client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java
+++ b/client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java
@@ -264,7 +264,7 @@ public class TcpClientConnection implements ClientConnection {
 
     @Override
     public void send(WireCommand cmd) throws ConnectionFailedException {
-        Exceptions.checkNotClosed(closed.get(), this);
+        checkIfClosed();
         try {
             encoder.write(cmd);
         } catch (IOException e) {
@@ -276,13 +276,19 @@ public class TcpClientConnection implements ClientConnection {
 
     @Override
     public void send(Append append) throws ConnectionFailedException {
-        Exceptions.checkNotClosed(closed.get(), this);
+        checkIfClosed();
         try {
             encoder.write(append);
         } catch (IOException e) {
             log.warn("Error writing to connection: {}", e.toString());
             close();
             throw new ConnectionFailedException(e);
+        }
+    }
+
+    private void checkIfClosed() throws ConnectionFailedException {
+        if (closed.get()) {
+            throw new ConnectionFailedException("Connection already closed");
         }
     }
 

--- a/client/src/test/java/io/pravega/client/connection/impl/ClientConnectionTest.java
+++ b/client/src/test/java/io/pravega/client/connection/impl/ClientConnectionTest.java
@@ -42,25 +42,25 @@ import static org.junit.Assert.assertTrue;
 public class ClientConnectionTest {
 
     private static class ReplyProcessor extends FailingReplyProcessor {
-        AtomicBoolean falure = new AtomicBoolean(false);
+        AtomicBoolean failure = new AtomicBoolean(false);
         List<Reply> replies = new Vector<>();
 
         @Override
         public void process(Reply reply) {
             if (reply.isFailure()) {
-                falure.set(true);
+                failure.set(true);
             }
             replies.add(reply);
         }
         
         @Override
         public void processingFailure(Exception error) {
-            falure.set(true);
+            failure.set(true);
         }
 
         @Override
         public void connectionDropped() {
-            falure.set(true);
+            failure.set(true);
         }
     }
 
@@ -88,7 +88,7 @@ public class ClientConnectionTest {
         clientConnection.send(new Append("segment", new UUID(1, 2), 1, new Event(Unpooled.EMPTY_BUFFER), 2));
         wireCommand = messages.take();
         assertEquals(WireCommandType.APPEND_BLOCK, wireCommand.getType());
-        assertFalse(processor.falure.get());
+        assertFalse(processor.failure.get());
     }
 
     @Test
@@ -106,7 +106,7 @@ public class ClientConnectionTest {
         clientConnection.send(new WireCommands.SetupAppend(1, new UUID(1, 2), "segment", ""));
         clientConnection.send(new Append("segment", new UUID(1, 2), 1, new Event(Unpooled.EMPTY_BUFFER), 2));
         server.sendReply(new WireCommands.AuthTokenCheckFailed(1, "Injected error"));
-        AssertExtensions.assertEventuallyEquals(true, () -> processor.falure.get(), 5000);
+        AssertExtensions.assertEventuallyEquals(true, () -> processor.failure.get(), 5000);
     }
 
     @Test
@@ -129,7 +129,7 @@ public class ClientConnectionTest {
             server.sendReply(new WireCommands.DataAppended(i, writerId, i, i - 1, i * 100));
         }
         AssertExtensions.assertEventuallyEquals(100, () -> processor.replies.size(), 5000);
-        assertFalse(processor.falure.get());
+        assertFalse(processor.failure.get());
     }
     
     @Test
@@ -155,6 +155,8 @@ public class ClientConnectionTest {
             }
         });
         assertTrue(clientConnection.toString(), clientConnection.isClosed());
+        AssertExtensions.assertThrows(ConnectionFailedException.class,
+                () -> clientConnection.send(new Append("segment", writerId, 100, new Event(Unpooled.wrappedBuffer(payload)), 1)));
     }
     
     @Test(timeout = 15000)
@@ -177,6 +179,8 @@ public class ClientConnectionTest {
             }
         });
         assertTrue(clientConnection.toString(), clientConnection.isClosed());
+        AssertExtensions.assertThrows(ConnectionFailedException.class, () -> clientConnection.send(new WireCommands.KeepAlive()));
+
     }
     
     @Test(timeout = 15000)
@@ -231,7 +235,7 @@ public class ClientConnectionTest {
         clientConnection.send(new WireCommands.SetupAppend(1, writerId, "segment", ""));
         clientConnection.send(new Append("segment", writerId, 1, new Event(Unpooled.wrappedBuffer(payload)), 1));
         server.sendReply(new WireCommands.DataAppended(1, writerId, 1, 0, 100));
-        AssertExtensions.assertEventuallyEquals(true, () -> processor.falure.get(), 5000);
+        AssertExtensions.assertEventuallyEquals(true, () -> processor.failure.get(), 5000);
     }
     
     @Test
@@ -262,7 +266,7 @@ public class ClientConnectionTest {
         assertEquals(AppendBlock.class, server.getReadCommands().take().getClass());
         assertEquals(AppendBlockEnd.class, server.getReadCommands().take().getClass());
         assertTrue(server.getReadCommands().isEmpty());
-        assertFalse(processor.falure.get());
+        assertFalse(processor.failure.get());
     }
 
 }

